walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,65 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Collections;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Comparator;
 
 public final class FindMeetingQuery {
+
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    ArrayList<Event> sortedEvents = new ArrayList<Event>(events);
+    Collection<TimeRange> freeTimeRanges = new ArrayList<TimeRange>();
+    Collection<TimeRange> freeTimeRangesWithOptionalAttendee = new ArrayList<TimeRange>();
+    int end = 0, FULL_DAY = 24*60, endMandatory = 0;
+    
+    Collections.sort(sortedEvents, new Comparator<Event>() {
+      @Override public int compare(Event e1, Event e2) {
+        return TimeRange.ORDER_BY_START.compare(e1.getWhen(), e2.getWhen());
+      }
+    });
+    
+    for(Event e: sortedEvents) {
+      if(!Collections.disjoint(e.getAttendees(), request.getAttendees())) {
+        if(e.getWhen().start() > end && e.getWhen().start()-request.getDuration() >= end) {
+          freeTimeRanges.add(TimeRange.fromStartEnd(end, e.getWhen().start(), false));
+        }
+
+        if(end < e.getWhen().end()) {
+          end = e.getWhen().end();
+        }
+      }
+    }
+
+    //If last event doesn't go to the end of the day
+    if(end < FULL_DAY && request.getDuration() < FULL_DAY && !request.getAttendees().isEmpty()) {
+      freeTimeRanges.add(TimeRange.fromStartEnd(end, FULL_DAY, false));
+    }
+
+    // Duplicated for optional
+    end = 0;
```

Comment:
You can add to freeTimeRangesWithOptionalAttendee the first time you iterate sortedEvents, so you only do one pass.